src/crypter.rs

```diff
@@ -0,0 +1,145 @@
+use aes_gcm::{
+    aead::{AeadCore, AeadInPlace, KeyInit, OsRng},
+    Aes128Gcm as AesGcm, Error, KeySizeUser, Nonce,
+};
+use base64::{engine::general_purpose::STANDARD, DecodeError, Engine};
+use std::{
+    collections::VecDeque,
+    fmt::{self, Debug, Formatter},
+    iter,
+    str::FromStr,
+    sync::Arc,
+};
+use typenum::marker_traits::Unsigned;
+
+pub type Key = aes_gcm::Key<AesGcm>;
+
+#[derive(Clone)]
+pub struct Crypter(Arc<CrypterInner>);
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum CrypterParseError {
+    #[error(transparent)]
+    Base64(#[from] DecodeError),
+
+    #[error("incorrect key length, must be {} bytes after base64 decode", <AesGcm as KeySizeUser>::KeySize::USIZE)]
+    IncorrectLength,
+
+    #[error("at least one key needed")]
+    Missing,
+}
+
+impl FromStr for Crypter {
+    type Err = CrypterParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut keys = s
+            .split(',')
+            .map(|s| {
+                STANDARD
+                    .decode(s)
+                    .map_err(CrypterParseError::Base64)
+                    .and_then(|v| Key::from_exact_iter(v).ok_or(CrypterParseError::IncorrectLength))
+            })
+            .collect::<Result<VecDeque<Key>, _>>()?;
+        let current_key = keys.pop_front().ok_or(CrypterParseError::Missing)?;
+        Ok(Self::new(current_key, keys))
+    }
+}
+
+impl Debug for Crypter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Crypter")
+            .field("current_key", &"..")
+            .field("past_keys", &self.0.past_keys.len())
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+struct CrypterInner {
+    current_key: Key,
+    past_keys: Vec<Key>,
+}
+
+impl From<Key> for Crypter {
+    fn from(key: Key) -> Self {
+        Self::new(key, [])
+    }
+}
+
+impl Crypter {
+    pub fn new(current_key: Key, past_keys: impl IntoIterator<Item = Key>) -> Self {
+        Self(Arc::new(CrypterInner {
+            current_key,
+            past_keys: past_keys.into_iter().collect(),
+        }))
+    }
+
+    pub fn random() -> Self {
+        Self::generate_key().into()
+    }
+
+    pub fn generate_key() -> Key {
```

Comment:
This method should ideally be attacked to `Key`, or if it's not worth a newtype, a free function.

Reply:
Why?

Reply:
Because generating a key has nothing to do with a Crypter.

Reply:
Leaving this as is because it mirrors the interface of [AesGcm::generate_key](https://docs.rs/aes-gcm/latest/aes_gcm/trait.KeyInit.html#method.generate_key)